### ISSUE TITLE
Target 'BeforeBuild' only processed if Rebuild is used

### ISSUE THREAD

**Issue by meastream** [Microsoft/PTVS#1241]
Hi,
I'm using PyQt5 and I want to generate python code from _*.ui_  files with `pyuic`. I add the _*.ui_ files and the generated _ui*.py_ files to the project.
I add the `CoreCompile`and `BeforeBuild` targets to my project file.

``` xml
<Target Name="CoreCompile" />
<Target Name="BeforeBuild">
    <Exec Command="(set PYTHONPATH=$(SearchPath)) &amp; python pyuic.py > pyuic.log"
          IgnoreExitCode="true"
          WorkingDirectory="$(MSBuildProjectDirectory)"/>
</Target>
```

Qt Creator is used to modify the ui files. The `BeforeBuild` target should be run if the ui files are modified or if they don't exist. But the `BeforeBuild` don't run if I start Debug (pressing F5) or if i build the project (pressing F7) after modifing the ui files. The only way to start the target is a rebuild of the project (pressing Ctrl+Alt+F7) or if the generated files don't exist.

Is this a bug in PTVS or a configuration error at the project file?

Thx.

Using:
PTVS 2.2.2
VS 2013 Pro U5

**Comment by zooba** [comment]
I know this answer is a long time coming, but I'll post it here in case other people find this page.

**tl;dr**: Change your file's build action to "Compile" or "Content"

In our [targets](https://github.com/Microsoft/PTVS/blob/master/Python/Product/BuildTasks/Microsoft.PythonTools.targets) file, we override the standard targets for providing VS with the list of source and output files from the project. Within VS, we will get these lists and compare the last modified times - if any sources are modified after the outputs, we will rebuild the project. You need to have imported this file in your project (which is required in recent versions of VS, but was not required in old versions).

[This line](https://github.com/Microsoft/PTVS/blob/master/Python/Product/BuildTasks/Microsoft.PythonTools.targets#L227) collects the source files, and you can see the item types we use: `Compile`, `Content` and `InputFiles`. The first two are settable through the Properties panel (generally F4) after selecting the file in Solution Explorer, while the last is a custom one for project-specific extensibility (that is, you can generate this item group yourself in a target that runs before `SourceFilesProjectOutputGroup`).

The easiest way to make this work is to select your file in Solution Explorer, press F4 (or whatever brings up the Properties panel), and change the Build Action to Compile or Content. This will then include it in our detection of whether your project needs to be rebuilt or not.

![image](https://user-images.githubusercontent.com/1693688/30869999-8a20acb2-a297-11e7-9650-b65e83683d92.png)

**Comment by meastream** [comment]
Hi zooba,

thx for the information.

Best regards,
meastream
